### PR TITLE
Add instances for rewriting in pairs.

### LINKED ISSUE #17060: RelationPairs should have instances for rewriting in pairs
The RelationPairs module of the Coq Standard Library lacks the following two instances, without which it is excruciatingly slow to do setoid rewriting within pairs.  These should be added.   
```
#[export] Instance proper_pair1: forall A B RA1 RA2 RB1 RB2 (RA : relation A) (RB : relation B),
    Proper (RA1 ==> RA2 ==> Basics.flip Basics.impl) RA
    -> Proper (RB1 ==> RB2 ==> Basics.flip Basics.impl) RB
    -> Proper (RA1 * RB1 ==> RA2 * RB2 ==> Basics.flip Basics.impl) (RA * RB)%signature.
Proof. cbv; intuition eauto. Qed.

#[export] Instance proper_pair2: forall A B RA1 RA2 RB1 RB2 (RA : relation A) (RB : relation B),
    Proper (RA1 ==> RA2 ==> Basics.impl) RA
    -> Proper (RB1 ==> RB2 ==> Basics.impl) RB
    -> Proper (RA1 * RB1 ==> RA2 * RB2 ==> Basics.impl) (RA * RB)%signature.
Proof. cbv; intuition eauto. Qed.
```
See also the discussion at https://coq.zulipchat.com/#narrow/stream/237977-Coq-users/topic/RelationPairs.20rewriting.20really.20slow

This proposal is endorsed by @JasonGross.   This issue applies to Coq 8.16, for example.

### PR DESCRIPTION
**Pull request by HaoYang670**
Fixes / closes #17060

<!-- Pointers to relevant developer documentation:

Contributing guide: https://github.com/coq/coq/blob/master/CONTRIBUTING.md

Test-suite: https://github.com/coq/coq/blob/master/test-suite/README.md

Changelog: https://github.com/coq/coq/blob/master/doc/changelog/README.md

Building the doc: https://github.com/coq/coq/blob/master/doc/README.md
Sphinx: https://github.com/coq/coq/blob/master/doc/sphinx/README.rst
doc_gram: https://github.com/coq/coq/blob/master/doc/tools/docgram/README.md

Overlays: https://github.com/coq/coq/blob/master/dev/ci/user-overlays/README.md

### CODE REVIEW

**Comment by HaoYang670** [comment]
Could someone please help to reivew this?

**Comment by Alizter** [comment]
@coqbot run full ci

**Comment by Alizter** [comment]
@coqbot bench

File: theories/Classes/RelationPairs.v
Comment (JasonGross) [inline theories/Classes/RelationPairs.v:168]: I would call this `Proper_RelProd_flip_impl` and the other one `Proper_RelProd_impl`.

Comment (HaoYang670) [reply theories/Classes/RelationPairs.v]: Good suggestion! Renamed them.

**Review by JasonGross — APPROVED** [review]
This looks good to me.  It might be worth it to also add an `iff` instance.  Given the Coq bench results and the fact that these are export rather than global, I think it's relatively unlikely that this will cause issues.

**Comment by HaoYang670** [comment]
Thank you for review @JasonGross.

>  It might be worth it to also add an iff instance.

Do you mean to prove 
```coq
#[export] Instance Proper_RelProd_flip_impl: forall A B RA1 RA2 RB1 RB2 (RA : relation A) (RB : relation B),
    (Proper (RA1 ==> RA2 ==> Basics.flip Basics.impl) RA /\ Proper (RB1 ==> RB2 ==> Basics.flip Basics.impl) RB)
    <-> Proper (RA1 * RB1 ==> RA2 * RB2 ==> Basics.flip Basics.impl) (RA * RB)%signature.
```
?

I got `iff is not a declared type class.` error when trying to instance it.

**Comment by JasonGross** [comment]
No, I mean something like
```coq
#[export] Instance Proper_RelProd_iff: forall A B RA1 RA2 RB1 RB2 (RA : relation A) (RB : relation B),
    Proper (RA1 ==> RA2 ==> iff) RA
    -> Proper (RB1 ==> RB2 ==> iff) RB
    -> Proper (RA1 * RB1 ==> RA2 * RB2 ==> iff) (RA * RB)%signature.
```

**Comment by HaoYang670** [comment]
Thank you, @JasonGross. I've added the `iff` instance. Please help to reivew.

File: theories/Classes/RelationPairs.v
Comment (JasonGross) [inline theories/Classes/RelationPairs.v:185]: Please don't use autogenerated names.  Is `firstorder eauto` enough to solve the goal?  If not, please name the variables explicitly with `intros <names>`

**Review by JasonGross — CHANGES_REQUESTED** [review]

File: theories/Classes/RelationPairs.v
Comment (HaoYang670) [reply theories/Classes/RelationPairs.v]: Unfortunately, only `firstorder eauto` cannot solve the goal. 
I have explicitly `intros` with names. Please review

**Review by JasonGross — APPROVED** [review]

**Comment by JasonGross** [comment]
@coqbot run full ci

**Comment by JasonGross** [comment]
@coqbot bench

**Comment by HaoYang670** [comment]
The bench failed.

**Comment by JasonGross** [comment]
Maybe we should just try again? @coqbot bench

**Comment by HaoYang670** [comment]
How should we analyze the bench result?

**Comment by Alizter** [comment]
@HaoYang670 There is no observable slowdown, everything is within noise range. The reason I asked for one was to check that we didn't cause any problems with hint databases causing longer build times.

**Comment by Alizter** [comment]
@JasonGross will you assign?

**Comment by JasonGross** [comment]
I can assign, but do we need reviews from @coq/typeclasses-maintainers and @coq/stdlib-maintainers ?

**Review by Alizter — APPROVED** [review]

**Comment by Alizter** [comment]
The PR has been open long enough with not much activity from those maintainers so I think that it is fine to go ahead with it.

**Comment by JasonGross** [comment]
@coqbot merge now

**Comment by Zimmi48** [comment]
FTR, it would have been preferable if these commits had been squashed together, and in particular to use a git rebase operation instead of a git merge to resolve conflicts.

**Comment by JasonGross** [comment]
Would it be possible to give coqbot a "squash and merge" option, as well as a "rebase and then force push to the branch" option?

**Comment by Zimmi48** [comment]
I don't think rebasing can be automated in general when there are conflicts, even if the conflicts were fixed in a merge commit. We could implement a coqbot rebase feature (there is an issue about this: https://github.com/coq/bot/issues/105), but it would only work when the conflicts can be auto-fixed or when there are no conflicts.

Regarding the "squash and merge" option, currently we require merge commits, so this would require first force-pushing the squashed commit to the PR branch, then merging the PR. I'm also a bit worried about contributors starting to abuse this and squashing too much. But this is doable indeed.

**Pull request merged**
